README.md
# HTU2XD

This is a Rust driver for the TE Connectivity Measurement Specialties HTU20D and HTU21D series I2C temperature and humidity sensors.

## License

0-clause BSD (see LICENSE.txt)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## v0.1.0 - 2021-04-18

Initial release with basic functionality


tests/i2c_mocked.rs
extern crate embedded_hal;
extern crate embedded_hal_mock;
extern crate htu2xd;
extern crate nb;

use std::io;

use embedded_hal_mock::i2c::{Mock, Transaction};
use embedded_hal_mock::MockError;
use htu2xd::{Htu2xd, Reading, Resolution, SupplyVoltage};

/// Address of the sensor
const ADDRESS: u8 = 0x40;

/// Reads the default values from the user register, changes all the options, and writes them back
#[test]
fn user_register() -> Result<(), Box<dyn std::error::Error>> {
    // This is the default value, but with the three reserved bits (3, 4, and 5) set to 1.
    // The same values for those bits must be written back.
    let default_register_value = 0b0011_1010;
    // Written-back register value: reserved bits are the same, all other bits are flipped
    let new_register_value = 0b1011_1101;
    let expected = [
        Transaction::write_read(ADDRESS, vec![0b1110_0111], vec![default_register_value]),
        Transaction::write(ADDRESS, vec![0b1110_0110, new_register_value]),
    ];
    let mut mock = Mock::new(&expected);

    let mut htu = Htu2xd::new();
    let mut register = htu.read_user_register(&mut mock)?;

    // Check that the correct value was read
    assert!(matches!(
        register.resolution(),
        Resolution::Humidity12Temperature14
    ));
    assert!(matches!(register.supply_voltage(), SupplyVoltage::High));
    assert_eq!(register.otp_reload_enabled(), false);
    assert_eq!(register.heater_enabled(), false);

    // Change everything that can be changed
    register.set_resolution(Resolution::Humidity11Temperature11);
    register.set_heater_enabled(true);
    register.set_otp_reload_enabled(true);
    // Check that things were changed
    assert!(matches!(
        register.resolution(),
        Resolution::Humidity11Temperature11
    ));
    assert_eq!(register.otp_reload_enabled(), true);
    assert_eq!(register.heater_enabled(), true);
    // Write changes back
    htu.write_user_register(&mut mock, register)?;

    mock.done();
    Ok(())
}

#[test]
fn temperature_humidity_clock_stretch() {
    let expected = [
        // Read temperature
        Transaction::write_read(ADDRESS, vec![0xe3], vec![0x4e, 0x85, 0x6b]),
        // Read humidity
        Transaction::write_read(ADDRESS, vec![0xe5], vec![0x68, 0x3a, 0x7c]),
    ];
    let mut mock = Mock::new(&expected);

    let mut htu = Htu2xd::new();
    match htu.read_temperature_blocking(&mut mock).unwrap() {
        Reading::Ok(reading) => {
            assert_eq!(reading.as_raw(), 0x4e84);
            let degrees_c = reading.as_degrees_celsius();
            assert!(degrees_c >= 7.04);
            assert!(degrees_c < 7.05);
        }
        Reading::ErrorLow => panic!("Unexpected error low"),
        Reading::ErrorHigh => panic!("Unexpected error high"),
    }
    match htu.read_humidity_blocking(&mut mock).unwrap() {
        Reading::Ok(reading) => {
            assert_eq!(reading.as_raw(), 0x6838);
            let percent = reading.as_percent_relative();
            assert!(percent >= 44.8);
            assert!(percent < 44.9);
        }
        Reading::ErrorLow => panic!("Unexpected error low"),
        Reading::ErrorHigh => panic!("Unexpected error high"),
    }

    mock.done();
}

#[test]
fn temperature_humidity_nak() {
    /// A ConnectionRefused error here represents a NAK
    fn is_nak(error: &MockError) -> bool {
        matches!(error, MockError::Io(io::ErrorKind::ConnectionRefused))
    }

    let expected = [
        // Start temperature read
        Transaction::write(ADDRESS, vec![0xf3]),
        // Several read attempts that return NAK
        Transaction::read(ADDRESS, vec![0u8; 3])
            .with_error(MockError::Io(io::ErrorKind::ConnectionRefused)),
        Transaction::read(ADDRESS, vec![0u8; 3])
            .with_error(MockError::Io(io::ErrorKind::ConnectionRefused)),
        Transaction::read(ADDRESS, vec![0u8; 3])
            .with_error(MockError::Io(io::ErrorKind::ConnectionRefused)),
        Transaction::read(ADDRESS, vec![0u8; 3])
            .with_error(MockError::Io(io::ErrorKind::ConnectionRefused)),
        // Done measuring, return results
        Transaction::read(ADDRESS, vec![0x4e, 0x85, 0x6b]),
        // Start humidity read
        Transaction::write(ADDRESS, vec![0xf5]),
        // Several read attempts that return NAK
        Transaction::read(ADDRESS, vec![0u8; 3])
            .with_error(MockError::Io(io::ErrorKind::ConnectionRefused)),
        Transaction::read(ADDRESS, vec![0u8; 3])
            .with_error(MockError::Io(io::ErrorKind::ConnectionRefused)),
        // Done measuring, return results
        Transaction::read(ADDRESS, vec![0x68, 0x3a, 0x7c]),
    ];
    let mut mock = Mock::new(&expected);

    let mut htu = Htu2xd::new();

    let mut temperature_step2 = htu.read_temperature(&mut mock).unwrap();
    // 4 NAK errors while the sensor is measuring
    for _ in 0..4 {
        let error = temperature_step2
            .read_result(&mut mock, is_nak)
            .unwrap_err();
        assert!(matches!(error, nb::Error::WouldBlock));
    }
    // Now the sensor is done
    match temperature_step2.read_result(&mut mock, is_nak).unwrap() {
        Reading::Ok(reading) => {
            assert_eq!(reading.as_raw(), 0x4e84);
            let degrees_c = reading.as_degrees_celsius();
            assert!(degrees_c >= 7.04);
            assert!(degrees_c < 7.05);
        }
        Reading::ErrorLow => panic!("Unexpected error low"),
        Reading::ErrorHigh => panic!("Unexpected error high"),
    }
    let mut humidity_step2 = htu.read_humidity(&mut mock).unwrap();
    // 2 NAK errors while the sensor is measuring
    for _ in 0..2 {
        let error = humidity_step2.read_result(&mut mock, is_nak).unwrap_err();
        assert!(matches!(error, nb::Error::WouldBlock));
    }
    // Now the sensor is done
    match humidity_step2.read_result(&mut mock, is_nak).unwrap() {
        Reading::Ok(reading) => {
            assert_eq!(reading.as_raw(), 0x6838);
            let percent = reading.as_percent_relative();
            assert!(percent >= 44.8);
            assert!(percent < 44.9);
        }
        Reading::ErrorLow => panic!("Unexpected error low"),
        Reading::ErrorHigh => panic!("Unexpected error high"),
    }

    mock.done();
}


src/lib.rs
#![no_std]

extern crate embedded_hal;
extern crate nb;

mod crc;
mod user_register;

pub use crate::user_register::{Resolution, SupplyVoltage, UserRegister};

use core::marker::PhantomData;
use core::slice;

use embedded_hal::blocking::i2c::{Read, Write, WriteRead};

use crate::crc::Crc;

/// Address of the sensor
const ADDRESS: u8 = 0x40;

mod sealed {
    pub trait SealedFromRaw {
        fn from_raw(raw: u16) -> Self;
    }
}
use self::sealed::SealedFromRaw;

/// HTU2XD driver that does not own the I2C bus
///
/// The type parameter I is the I2C bus. This prevents one Htu2xd object from accidentally being
/// used with two different I2C peripherals.
///
/// # I2C type requirements
///
/// The `I` I2C bus type must return the same error type for read, write, and read/write operations.
///
/// # Examples
///
/// ## Configuration
///
/// ```no_run
/// use embedded_hal::blocking::delay::DelayMs;
/// use embedded_hal::blocking::i2c::{Read, Write, WriteRead};
/// use htu2xd::{Htu2xd, Resolution};
///
/// fn init_htu2xd<I, E, D>(i2c: &mut I, delay: &mut D) -> Result<Htu2xd<I>, E>
/// where
///     I: Read<Error = E> + Write<Error = E> + WriteRead<Error = E>,
///     D: DelayMs<u32>,
/// {
///     let mut htu = Htu2xd::new();
///
///     htu.soft_reset(i2c)?;
///     // Wait for the reset to finish
///     delay.delay_ms(15u32);
///
///     let mut register = htu.read_user_register(i2c)?;
///     register.set_resolution(Resolution::Humidity10Temperature13);
///     htu.write_user_register(i2c, register)?;
///
///     Ok(htu)
/// }
/// ```
///
/// ## Basic operation
///
/// ```no_run
/// use embedded_hal::blocking::i2c::{Read, Write, WriteRead};
/// use htu2xd::{Htu2xd, Reading, Temperature};
/// fn use_sensor<I, E>(htu: &mut Htu2xd<I>, i2c: &mut I) -> Result<(), htu2xd::Error<E>>
/// where
///     I: Write<Error = E> + Read<Error = E> + WriteRead<Error = E>,
/// {
///     let temperature_reading = htu.read_temperature_blocking(i2c)?;
///     match temperature_reading {
///         Reading::Ok(reading) => {
///             println!("Temperature {} degrees C", reading.as_degrees_celsius())
///         }
///         Reading::ErrorLow => println!("Temperature off-scale low or sensor error"),
///         Reading::ErrorHigh => println!("Temperature off-scale high or sensor error"),
///     }
///     let humidity_reading = htu.read_humidity_blocking(i2c)?;
///     match humidity_reading {
///         Reading::Ok(reading) => println!("Humidity {}%", reading.as_percent_relative()),
///         Reading::ErrorLow => println!("Humidity off-scale low or sensor error"),
///         Reading::ErrorHigh => println!("Humidity off-scale high or sensor error"),
///     }
///     Ok(())
/// }
/// ```
///
/// ## Temperature and humidity reading without clock stretching
///
/// ```no_run
/// use embedded_hal::blocking::i2c::{Read, Write, WriteRead};
/// use htu2xd::{Htu2xd, Reading, Temperature};
///
/// enum I2cError {
///     Nak,
///     OtherError,
/// }
///
/// impl I2cError {
///     fn is_nak(&self) -> bool {
///         matches!(self, I2cError::Nak)
///     }
/// }
///
/// fn use_sensor<I>(htu: &mut Htu2xd<I>, i2c: &mut I) -> Result<(), htu2xd::Error<I2cError>>
/// where
///     I: Write<Error = I2cError> + Read<Error = I2cError> + WriteRead<Error = I2cError>,
/// {
///     let mut temperature_step2 = htu.read_temperature(i2c)?;
///     // Do something else while the sensor is busy
///     // Later, read the result
///     let temperature_reading = nb::block!(temperature_step2.read_result(i2c, I2cError::is_nak))?;
///     match temperature_reading {
///         Reading::Ok(reading) => {
///             println!("Temperature {} degrees C", reading.as_degrees_celsius())
///         }
///         Reading::ErrorLow => println!("Temperature off-scale low or sensor error"),
///         Reading::ErrorHigh => println!("Temperature off-scale high or sensor error"),
///     }
///     let mut humidity_step2 = htu.read_humidity(i2c)?;
///     // Do something else while the sensor is busy
///     // Later, read the result
///     let humidity_reading = nb::block!(humidity_step2.read_result(i2c, I2cError::is_nak))?;
///     match humidity_reading {
///         Reading::Ok(reading) => println!("Humidity {}%", reading.as_percent_relative()),
///         Reading::ErrorLow => println!("Humidity off-scale low or sensor error"),
///         Reading::ErrorHigh => println!("Humidity off-scale high or sensor error"),
///     }
///     Ok(())
/// }
/// ```
pub struct Htu2xd<I>(PhantomData<I>);

impl<I, E> Htu2xd<I>
where
    I: Read<Error = E> + Write<Error = E> + WriteRead<Error = E>,
{
    /// Creates a driver object, but does not perform any initialization
    pub fn new() -> Self {
        Htu2xd(PhantomData)
    }

    /// Resets the sensor and restores default settings, but does not restore the heater enable bit
    ///
    /// After this function returns the sensor may take up to 15 ms to reset.
    pub fn soft_reset(&mut self, i2c: &mut I) -> Result<(), E> {
        i2c.write(ADDRESS, &[Command::SoftReset as u8])
    }

    /// Reads the current humidity
    ///
    /// In this mode, the sensor stretches the I2C clock while it takes a measurement. This
    /// function blocks until the measurement has finished and been read.
    pub fn read_humidity_blocking(&mut self, i2c: &mut I) -> Result<Reading<Humidity>, Error<E>> {
        let mut buffer = [0u8; 3];
        i2c.write_read(ADDRESS, &[Command::HumidityHoldMaster as u8], &mut buffer)?;
        parse_and_check_reading(&buffer)
    }

    /// Reads the current temperature
    ///
    /// In this mode, the sensor stretches the I2C clock while it takes a measurement. This
    /// function blocks until the measurement has finished and been read.
    pub fn read_temperature_blocking(
        &mut self,
        i2c: &mut I,
    ) -> Result<Reading<Temperature>, Error<E>> {
        let mut buffer = [0u8; 3];
        i2c.write_read(
            ADDRESS,
            &[Command::TemperatureHoldMaster as u8],
            &mut buffer,
        )?;
        parse_and_check_reading(&buffer)
    }

    /// Reads the current humidity
    ///
    /// In this mode, the sensor does not stretch the I2C clock. After sending the command to
    /// the sensor, this function returns a proxy that can be polled to determine if the result
    /// is ready.
    pub fn read_humidity(&mut self, i2c: &mut I) -> Result<ResultReader<I, Humidity>, E> {
        // Send a command to start the read
        i2c.write(ADDRESS, &[Command::Humidity as u8])?;
        Ok(ResultReader {
            _driver: PhantomData,
            _reading: PhantomData,
        })
    }

    /// Reads the current temperature
    ///
    /// In this mode, the sensor does not stretch the I2C clock. After sending the command to
    /// the sensor, this function returns a proxy that can be polled to determine if the result
    /// is ready.
    pub fn read_temperature(&mut self, i2c: &mut I) -> Result<ResultReader<I, Temperature>, E> {
        // Send a command to start the read
        i2c.write(ADDRESS, &[Command::Temperature as u8])?;
        Ok(ResultReader {
            _driver: PhantomData,
            _reading: PhantomData,
        })
    }

    /// Reads the user register and returns its content
    pub fn read_user_register(&mut self, i2c: &mut I) -> Result<UserRegister, E> {
        let mut register_value = 0u8;
        i2c.write_read(
            ADDRESS,
            &[Command::ReadUser as u8],
            slice::from_mut(&mut register_value),
        )?;
        Ok(UserRegister(register_value))
    }

    /// Writes the user register
    ///
    /// You must use the `read_user_register` function to get a `UserRegister` object that
    /// can be modified and then passed to this function.
    pub fn write_user_register(&mut self, i2c: &mut I, register: UserRegister) -> Result<(), E> {
        i2c.write(ADDRESS, &[Command::WriteUser as u8, register.0])
    }
}

impl<I, E> Default for Htu2xd<I>
where
    I: Read<Error = E> + Write<Error = E> + WriteRead<Error = E>,
{
    fn default() -> Self {
        Htu2xd::new()
    }
}

/// A proxy used to read the result of a non-blocking measurement
pub struct ResultReader<'h, I, M> {
    _driver: PhantomData<&'h mut Htu2xd<I>>,
    _reading: PhantomData<M>,
}

impl<'h, I, M> ResultReader<'h, I, M>
where
    I: Read,
    M: Measurement,
{
    /// Attempts to read a measurement result from the sensor
    ///
    /// is_nak must be a closure that returns true if the provided error is a NAK (negative
    /// acknowledge) error, or false otherwise.
    ///
    /// This function returns `Err(nb::Error::WouldBlock)` if the sensor does not acknowledge
    /// its address. This means that it is still performing the measurement. This function should
    /// be called again later to try again.
    ///
    /// On success, this function returns the sensor reading.
    ///
    /// After this function returns anything other than `Err(nb::Error::WouldBlock)`, this
    /// `ResultReader` must not be used again.
    pub fn read_result<F>(
        &mut self,
        i2c: &mut I,
        is_nak: F,
    ) -> nb::Result<Reading<M>, Error<I::Error>>
    where
        F: FnOnce(&I::Error) -> bool,
    {
        let mut buffer = [0u8; 3];
        match i2c.read(ADDRESS, &mut buffer[..]) {
            Ok(()) => parse_and_check_reading(&buffer).map_err(nb::Error::Other),
            Err(e) => {
                if is_nak(&e) {
                    // Measurement is still in progress, try again later
                    Err(nb::Error::WouldBlock)
                } else {
                    // Some other error, not a NAK
                    Err(nb::Error::Other(Error::I2c(e)))
                }
            }
        }
    }
}

/// Checks the CRC of a 3-byte temperature or humidity reading and parses it as a `Reading` object
fn parse_and_check_reading<M, E>(bytes: &[u8; 3]) -> Result<Reading<M>, Error<E>>
where
    M: Measurement,
{
    // Check CRC
    let mut crc = Crc::new();
    crc.add_all(&*bytes);
    if crc.value() != 0 {
        return Err(Error::Crc);
    }

    // Parse reading
    let reading16 = (u16::from(bytes[0]) << 8) | u16::from(bytes[1]);

    Ok(Reading::from_raw(reading16))
}

/// An I2C or CRC error
#[derive(Debug)]
pub enum Error<E> {
    /// The I2C driver returned an error
    I2c(E),
    /// A message was received from the sensor with an invalid CRC checksum
    Crc,
}

impl<E> From<E> for Error<E> {
    fn from(inner: E) -> Self {
        Error::I2c(inner)
    }
}

/// A temperature reading
#[derive(Debug, Clone)]
pub struct Temperature(u16);

impl Temperature {
    /// Returns the temperature reading exactly as read from the sensor, with the status bits
    /// cleared
    pub fn as_raw(&self) -> u16 {
        self.0
    }

    /// Converts the temperature reading into degrees Celsius
    ///
    /// This function uses single-precision floating-point operations.
    pub fn as_degrees_celsius(&self) -> f32 {
        -46.85_f32 + 175.72_f32 / 65536.0_f32 * f32::from(self.0)
    }
}

/// A humidity reading
#[derive(Debug, Clone)]
pub struct Humidity(u16);

impl Humidity {
    /// Returns the humidity reading exactly as read from the sensor, with the status bits cleared
    pub fn as_raw(&self) -> u16 {
        self.0
    }

    /// Converts the temperature reading into percent relative humidity (0.0 = 0%, 100.0 = 100%)
    /// and clamps it to the 0%-100% range
    ///
    /// This function uses single-precision floating-point operations.
    pub fn as_percent_relative(&self) -> f32 {
        -6.0_f32 + 125.0_f32 / 65536.0_f32 * f32::from(self.0)
    }
}

pub trait Measurement: SealedFromRaw {}
impl SealedFromRaw for Temperature {
    fn from_raw(raw: u16) -> Self {
        Temperature(raw)
    }
}
impl Measurement for Temperature {}
impl SealedFromRaw for Humidity {
    fn from_raw(raw: u16) -> Self {
        Humidity(raw)
    }
}
impl Measurement for Humidity {}

/// Information about a temperature or humidity reading
#[derive(Debug, Clone)]
pub enum Reading<R> {
    /// The reading was completed normally
    Ok(R),
    /// The reading was very low, or the sensor has an open circuit
    ErrorLow,
    /// The reading was very high, or the sensor has a short circuit
    ErrorHigh,
}

impl<R> Reading<R>
where
    R: Measurement,
{
    fn from_raw(raw: u16) -> Self {
        match raw {
            0x0000 => Reading::ErrorLow,
            0xffff => Reading::ErrorHigh,
            _ => {
                // Clear the status bits (lowest two) and return the reading
                Reading::Ok(R::from_raw(raw & 0xfffc))
            }
        }
    }
}

/// Commands to read and write things
enum Command {
    TemperatureHoldMaster = 0xe3,
    Temperature = 0xf3,
    HumidityHoldMaster = 0xe5,
    Humidity = 0xf5,
    WriteUser = 0xe6,
    ReadUser = 0xe7,
    SoftReset = 0xfe,
}


src/user_register.rs
//! Definitions for the user register

/// Resolution for temperature and humidity measurements
///
/// Lower resolutions take less time to measure.
#[derive(Debug, Clone)]
pub enum Resolution {
    /// 12-bit humidity, 14-bit temperature
    Humidity12Temperature14,
    /// 8-bit humidity, 12-bit temperature
    Humidity8Temperature12,
    /// 10-bit humidity, 13-bit temperature
    Humidity10Temperature13,
    /// 11-bit humidity, 11-bit temperature
    Humidity11Temperature11,
}

/// Measurement of the power supply voltage
///
/// Note: The sensor's minimum power supply voltage is 1.5 V.
#[derive(Debug, Clone)]
pub enum SupplyVoltage {
    /// Greater than 2.25 +/- 0.1 V
    High,
    /// Less than 2.25 +/- 0.1 V
    Low,
}

/// The user register, used for configuration
///
/// The only way to create a `UserRegister` object is to read it from a sensor. It can then be
/// modified and written back.
// The enclosed value is represented exactly as the sensor sends and receives it.
pub struct UserRegister(pub(crate) u8);

impl UserRegister {
    /// Returns the current measurement resolution
    pub fn resolution(&self) -> Resolution {
        let bit7 = ((self.0 >> 7) & 1) == 1;
        let bit0 = (self.0 & 1) == 1;
        match (bit7, bit0) {
            (false, false) => Resolution::Humidity12Temperature14,
            (false, true) => Resolution::Humidity8Temperature12,
            (true, false) => Resolution::Humidity10Temperature13,
            (true, true) => Resolution::Humidity11Temperature11,
        }
    }
    /// Returns the supply voltage when the last temperature or humidity measurement was taken
    pub fn supply_voltage(&self) -> SupplyVoltage {
        let bit6 = ((self.0 >> 6) & 1) == 1;
        if bit6 {
            SupplyVoltage::Low
        } else {
            SupplyVoltage::High
        }
    }
    /// Returns true if the on-chip heater is enabled
    pub fn heater_enabled(&self) -> bool {
        ((self.0 >> 2) & 1) == 1
    }
    /// Returns true if the one-time programmable memory reload is active
    ///
    /// With this active, the default settings will be restored after each temperature or
    /// humidity measurement.
    pub fn otp_reload_enabled(&self) -> bool {
        let bit1 = ((self.0 >> 1) & 1) == 1;
        // Invert (1 = disable)
        !bit1
    }

    ///  Sets the measurement resolution
    pub fn set_resolution(&mut self, resolution: Resolution) {
        let bit_7_mask = 1u8 << 7;
        let bit_0_mask = 1u8;
        // Clear bits 7 and 0
        self.0 &= !(bit_7_mask | bit_0_mask);
        match resolution {
            Resolution::Humidity12Temperature14 => { /* Leave both bits clear */ }
            Resolution::Humidity8Temperature12 => {
                // Set bit 0
                self.0 |= bit_0_mask;
            }
            Resolution::Humidity10Temperature13 => {
                // Set bit 7
                self.0 |= bit_7_mask;
            }
            Resolution::Humidity11Temperature11 => {
                // Set bits 7 and 0
                self.0 |= bit_7_mask | bit_0_mask;
            }
        }
    }
    /// Enables or disables the on-chip heater
    pub fn set_heater_enabled(&mut self, enabled: bool) {
        let bit_2_mask = 1u8 << 2;
        if enabled {
            self.0 |= bit_2_mask;
        } else {
            self.0 &= !bit_2_mask;
        }
    }
    /// Enables or disables the reloading of default settings from one-time programmable memory
    /// after each measurement
    pub fn set_otp_reload_enabled(&mut self, otp_reload: bool) {
        let bit_1_mask = 1u8 << 1;
        if otp_reload {
            // Clear bit (0 = enable)
            self.0 &= !bit_1_mask;
        } else {
            // Set bit (1 = disable)
            self.0 |= bit_1_mask;
        }
    }
}

mod debug_impl {
    use super::UserRegister;
    use core::fmt::{Debug, Formatter, Result};

    impl Debug for UserRegister {
        fn fmt(&self, f: &mut Formatter<'_>) -> Result {
            f.debug_struct("UserRegister")
                .field("resolution", &self.resolution())
                .field("supply_voltage", &self.supply_voltage())
                .field("heater_enabled", &self.heater_enabled())
                .field("otp_reload_enabled", &self.otp_reload_enabled())
                .finish()
        }
    }
}


src/crc.rs
//! CRC for communication integrity checks

/// 8-bit CRC, polynomial x^8 + x^5 + x^4 + 1, initialization 0
pub struct Crc {
    value: u8,
}

impl Crc {
    /// Creates a new CRC calculator initialized to zero
    pub fn new() -> Self {
        Crc { value: 0 }
    }

    /// Adds a byte to the CRC
    pub fn add(&mut self, byte: u8) {
        self.value ^= byte;
        for _ in 0..8 {
            if (self.value & 0x80) != 0 {
                self.value = (self.value << 1) ^ 0x31;
            } else {
                self.value <<= 1;
            }
        }
    }
    /// Adds bytes from a slice
    pub fn add_all(&mut self, bytes: &[u8]) {
        for &byte in bytes {
            self.add(byte);
        }
    }
    /// Returns the current remainder value
    pub fn value(&self) -> u8 {
        self.value
    }
}

#[cfg(test)]
mod crc_test {
    use super::Crc;
    #[test]
    fn short() {
        let mut crc = Crc::new();
        crc.add(0xdc);
        assert_eq!(crc.value(), 0x79);
    }

    #[test]
    fn message_a() {
        let mut crc = Crc::new();
        crc.add_all(&[0x68, 0x3a]);
        assert_eq!(crc.value(), 0x7c);
    }

    #[test]
    fn message_b() {
        let mut crc = Crc::new();
        crc.add_all(&[0x4e, 0x85]);
        assert_eq!(crc.value(), 0x6b);
    }
}
